### ISSUE TITLE
Prioritization of access restrictions

### ISSUE THREAD

**Issue by RastislavKiss** [sandboxie/sandboxie#13]
Hi there,

## Problem

a real-world situation, I have an unknown app, which I have installed. I want to use it, but I want to take some security measures.
In %appdata% is configuration of Thunderbird, with all my mails in a database. Similarly, there is Firefox configuration with saved passwords and other sensitive stuff. I don't want the app to be able to read them, so I set my sandbox to completely block %appdata%.
But now, there is a problem, the app can't save its own settings. Because the blocking rule takes precedence, I can't do anything about it, just block every single folder of %appdata%, leaving only the one of the app. That's not just quite exhaustive, but also inefficient, as I'll need to continually add new folders of other programs to the block list, and also quite errorprone, as I could forget to mark one of the folders.
This problem infact targets much wider spectrum, in general I would like to prevent unknown apps from accessing any irrelevant location of the system, allowing just specific parts to be accessed. It's like when you get a visit, the visitors should always stay with you and max. go to the toilet, you usually don't let unknown people to freely walk around your house, like any app can read almost anything in Windows.
But I took just the appdata example, to keep it simple.

## Solution

Solution imo is quite simple. Instead of setting file access rules individually, they could be set in a list. During adding a rule, user could select, what access type is the rule of and to which location is it relevant. 
Then by moving items up and down in the list, one could very accurately specify how exactly should overlap not just allow / block rules, but also other kinds. It would be far simpler to create for example a completely readonly area with just one modifiable folder inside, or wiceversa.
Each program could have such a list, like it's now possible to specify rules just for certain programs.
It would be not just straight-forward and very intuitive, but also computationally cheap. Sandboxie would need to do just two things when file access event occurs:

1. Check, if current sandbox has a specific list for particular program.
2. narrow down the list and apply the first rule relevant to the accessed location.

That's all. The same idea could be applied also for registry, as they work similarly to the filesystem tree structure.
I think, that this form would be much more flexible, than the current messy layout.

**Comment by Deltrego** [comment]
You can already configure a folder for Write-Only Access (it appears empty to sandboxed programs but they can write their own files and read them) ... would that be enough? Don't use Blocked Access.
To test it, choose your sandbox, then "Run Windows Explorer", attempt to go to %APPDATA% with your settings and see if you can access any files.

**Comment by DavidXanatos** [comment]
@RastislavKiss you are right that the non existence of data protection is an issue.
I think the best solution is to redirect all sandboxed requests by default to a new empty user profile.

So when creating a sanbox you could pick a normal one as it behaves now, 
or a "private" or "anonymous" one that does not expose any user data to the running programs and ideally hides even the the user name.

I'm looking forward to adding a functionality like that to my fork.

**Comment by RastislavKiss** [comment]
Hi there,
@Deltrego: yes and no. It would work, upto point, where I would want the sandboxed app to have access to certain directory in appdata. Writeonly and readonly rules would then overlap, what is not possible to solve with current way of setting these rules in Sandboxie.
It's not an unreal situation, Python for example is located in appdata by default. If I wanted to sandbox a Python code editor, I don't want it to read my mails, so I would set writeonly for appdata, but by that step, I would also cut off Python, what means lost compiling, running and debugging functionality of the ide.
On the other side, if these access rules were managed in one list per program, or for all programs, it would be trivial to set readonly rule for python directory as first, and writeonly rule for appdata as second.
So whenever the ide wanted to read something from Python's directory, checking the list from top to bottom would hit the readonly rule first, and grant readonly access for that particular action, whereas if the ide wanted to spy something from Thunderbird directory, that doesn't fall under the first readonly rule, so the secondone would match and the curious program would be out of luck.

There are of course more programs of this kind, for example sandboxing radare's ide would result in the same problem. And those are just general situations, someone might have even more specific needs, which would be totally unfillable, without a way to set priorities for individual access rules.

@DavidXanatos: that might be a possible solution, but it would make moving dependencies such as Python into the empty user's sandbox quite a drag. And also if you wanted to separate dependencies for various programs, that would require creating unique sandbox for each situation, what would be drag as well. :)
Although a drag is a quite subjective measure, I think, that the list way of managing rules would do the job equally well, but without need to move things around and with much easier implementation and setting up.

**Comment by DavidXanatos** [comment]
Yes, a list of allow/deny rules that can be prioritized would definitely be an improvement and needed in some use cases.

The issue here however is that yous till need to set up all the required block rules to protect your privacy in the first place.

So may be a good approach would be to have an empty user account into which selected folders and reg keys from the real account could be mapped into.
